### ISSUE TITLE
Rename this module

### ISSUE THREAD

**Issue by mcollina** [fastify/fastify-plugin#207]
From a discussing with @lirantal on twitter: https://twitter.com/matteocollina/status/1610577242844708865?s=46&t=wpaDKCujckm07OZ7JAa3hg:

> Unfortunately there is. I made some terminology mistakes at the beginning and they stuck, eg “everything is a plugin”. But also “plugins are encapsulated by default”. “fastify-plugin” is for creating reusable modules, not for application builders.

I think we should rename this module to `@fastify/meta`.

**Comment by delvedor** [comment]
I agree with what @jsumners said. Yes, in retrospect, it's a bad name, but changing it now could cause more problems than what it would actually solve. If we decide to rename it, it should be a name that is so good that it's impossible to say no :P

Some ideas:
- `@fastify/share-plugin`
- `@fastify/decontextualize`
- `@fastify/plugin-maintainer-essentials`

**Comment by fox1t** [comment]
I agree with @jsumners and @delvedor, but, as I said on that Twitter thread, it is impossible for me to unsee the issue with that name. 
I want to give it a shot: we can use the "opposite" of encapsulate.
- `fastify-expose` -> `@fastify/expose-plugin`

Self-explanatory enough.

**Comment by Eomm** [comment]
I would add to the table:

- `@fastify/sidecar-metadata`: it is an optional way to set some metadata on a plugin

----

I think after collecting some ideas, we could just vote the proposed option 🔢 / or to not rename it 👎🏽

**Comment by climba03003** [comment]
Beware of we provide a option for `encapsulate`, so it by default exclude from encapsulation but also allowed to create encapsulation.

**Comment by Uzlopak** [comment]
I had a discussion a week ago wth liran tal about that issue in linkedin.
https://www.linkedin.com/posts/talliran_hey-fastifyjs-devs-a-thought-would-it-activity-7014228764804952064-bcKB?utm_source=share&utm_medium=member_desktop

I think the name, fastify-plugin, is good.

**Comment by climba03003** [comment]
> I would avoid this, encapsulation by default is a design choice.

Maybe you somehow mistaken? It is a option in this package.
https://github.com/fastify/fastify-plugin#encapsulate

So, any name that imply it must escape the encapsulation will be wrong because you have ability to control the behavior.

**Comment by jsumners** [comment]
Here's the thing, "plugin" has two simultaneous definitions, with one being a subset of the other:

1. a simple function that encapsulates logic (e.g. other plugins and routes)
2. a simple function that encapsulates logic in a distributable package with corresponding metadata (i.e. `fastify-plugin` has been applied to it)

The fact that a plugin can have its context exposed to ancestor contexts or not is not really the point.

**Comment by mcollina** [comment]
@jsumners +1. That's pretty much the sum of the discussion. This is mostly a utility module so that all the metadata needed by Fastify can be set in a convenient way instead of using global symbols.

(This is not part of Fastify to avoid individual packages to depend on Fastify itself.)

The question is how we fix this. It's confusing for most people, and I think we should tame this problem.
We do have a slot for releasing a Fastify v5 release next year (to drop both v14 and v16), so we might think of a breaking change as well.

**Comment by mcollina** [comment]
`@fastify/annotate`?

**Comment by jsumners** [comment]
> `@fastify/annotate`?

But what does it annotate?

**Comment by mcollina** [comment]
Plugins!

**Comment by Uzlopak** [comment]
Tomorrow we will not only annotate some symbols but encapsulate some additional checks and then we need a new npm package covering that behaviour.

**Comment by jsumners** [comment]
> Plugins!

I know, but the name does not make that clear. When people are looking through our list of modules, the name of those modules should give some indication of their purpose.

**Comment by mcollina** [comment]
How about `@fastify/add-metadata`?

**Comment by fox1t** [comment]
What about `@fastify/plugin-metadata`? So we can conserve the plugin part in the name (which I agree with @jsumners is important) and we say what the package does (manipulates the plugin metadata).

**Comment by Uzlopak** [comment]
fastify-module.

**Comment by d1b1** [comment]
As a newbie to fastify and its plugin system, the naming conventions are lead into how to think about packages. Given that plugins carry its own baggage, a clear README that makes sure developers can read and understand the fastify approach might be the best solution. 

'@fastify/plugin-expose' is the closest, and would make me dig into understand the plugin framework in fastify

Maybe '@fastify/plugin-extras' ?

**Comment by lirantal** [comment]
Is `@fastify/plugin-export` not a better alternative @d1b1 ?
As a module author, I don't see anything about the naming `plugin-export` that would peak my curiosity to further investigate what it does under the hood. It's superior in that regards to _expose_ which leaks a potential implementation detail where-as _export_ is quite neutral.

**Comment by d1b1** [comment]
Gotcha, in the end the name is short enough, that anyone who cares has to go into the Readme. So yes, 'export' does the same thing, and better. Export is a node term as well, so leads the dev to thinks node vs export data.

**Comment by lirantal** [comment]
_export_ is also an ESM term so it works for everyone in JavaScript :-)

**Comment by mcollina** [comment]
I think `@fastify/export` is nice and short. Wdyt folks?

**Comment by jsumners** [comment]
Someone (@mcollina) should put out a poll with the options:

* `@fastify/add-metadata`
* `@fastify/plugin-metadata`
* `@fastify/plugin-export`
* `@fastify/export`

Let the wider community decide (since they are very likely not reading this thread).

After that, the question becomes who will take on the work.

**Comment by jsumners** [comment]
> What should I use to create a poll?

Not sure. Does the OpenJSF have a tool we can use?

> Happy to promote it on my social media.

Exactly. You have the most reach for such a poll.

**Comment by Joabesv** [comment]
Hi guys, just wanted to know if the module rename will happen? or it got put aside

**Comment by mcollina** [comment]
It _will_ happen. I've been sidetracked by various things. At this rate, it might happen when we ship Fastify v5.

**Comment by gurgunday** [comment]
I vote for `@fastify/export` 😁
